sb3_contrib/common/maskable/buffers.py

```diff
@@ -0,0 +1,85 @@
+from typing import Generator, NamedTuple, Optional
+
+import numpy as np
+import torch as th
+from gym import spaces
+from stable_baselines3.common.buffers import RolloutBuffer
+from stable_baselines3.common.vec_env import VecNormalize
+
+
+# TODO use dataclass when Python 3.6 support is dropped
+class MaskedRolloutBufferSamples(NamedTuple):
+    observations: th.Tensor
+    actions: th.Tensor
+    old_values: th.Tensor
+    old_log_prob: th.Tensor
+    advantages: th.Tensor
+    returns: th.Tensor
+    action_masks: th.Tensor
+
+
+class MaskedRolloutBuffer(RolloutBuffer):
+    def __init__(self, *args, **kwargs):
+        self.action_masks = None
+        super().__init__(*args, **kwargs)
+
+    def reset(self) -> None:
+        # TODO: need to handle different action_space attrs based on shape
+        action_masks = np.array([], dtype=np.float32)
+        if isinstance(self.action_space, spaces.Discrete):
+            action_masks = np.ones((self.buffer_size, self.n_envs, self.action_space.n), dtype=np.float32)
+        elif isinstance(self.action_space, spaces.MultiDiscrete):
+            action_masks = np.ones(
+                (self.buffer_size, self.n_envs, sum(self.action_space.nvec)),
+                dtype=np.float32,
+            )
+        self.action_masks = action_masks
+
+        super().reset()
+
+    def add(self, *args, action_masks: np.ndarray = None, **kwargs) -> None:
+        """
+        :param action_masks: Masks applied to constrain the choice of possible actions.
+        """
+        if action_masks is not None:
+            self.action_masks[self.pos] = action_masks.flatten()
```

Comment:
@benallard can you confirm that this change works for you? I wasn't able to use your suggestion because not all gym action spaces have the `nvec` attribute, it's specfic to multidiscrete.

Reply:
not sure what was the issue (what is the expected shape?)... maybe you can take a look at https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/buffers.py#L364

and 
https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/buffers.py#L45 defined https://github.com/DLR-RM/stable-baselines3/blob/ddbe0e93f9fe55152f2354afd058b28e6ccc3345/stable_baselines3/common/preprocessing.py#L161

but I assume you are expecting the action mask to be already preprocessed? (one hot vector)

Reply:
The issue was that the code originally expected the mask array to be a 1D array, regardless of the shape of the action space:
```
        if action_masks is not None:
            self.action_masks[self.pos] = np.array(action_masks).copy()
```
This was a surprise to benallard. He tried to pass a mask with a separate array per choice in his multidiscrete action space:
```
def mask_fn(env: gym.Env) -> np.ndarray:
    positions = set(a[0] for a in env.legal_actions)
    pieces = set(a[1] for a in env.legal_actions)
    action_mask = [[], []]  # Separate array per choice in multidiscrete
    for i in range(16):
        action_mask[0].append(i in positions)
        action_mask[1].append(i in pieces)
    return action_mask
```

I thought it would be better not to be strict about the shape of the provided mask and simply flatten it internally. We could manually reshape to match our expectation, per benallard's earlier suggestion:
```
self.action_masks[self.pos] = np.array(action_masks).reshape(1, sum(self.action_space.nvec)).copy()
```
The approach for doing this would differ based the kind of action space used. I thought it wasn't necessary because the mask is reshaped again [when it's applied to the distribution](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/25/files#diff-6a50138bae28ed7a66f6feda0989635797e3ee4e5adaa0d78ebc63b8da5ada3bR180-R184). If the shape is wrong, it would be caught then.

Reply:
I can confirm that it works 👍